### PR TITLE
feat: update to new libp2p interfaces

### PR DESCRIPTION
**Pull request by achingbrain**
Updates this module to use the `@libp2p/*` set of modules as part of the typescript port.

- Renames imports to add .js suffix (tsc does not add this during compilation to ESM)
- Renames bundled `@types/*.d.ts` to `@types/*.ts` because tsc ignores `.d.ts` which masks errors
- Updates project config to output ESM instead of CJS code
- Updates all deps to lastest versions including ones that have gone ESM-only
- Re-generates protobuf code with the ES6 target
- Fixes all linting errors
- Swaps `bl` for `Uint8ArrayList`

N.b. `pbjs` generates invalid imports for bundler-less ESM - the `.js` suffix has to be added to the import manually so it's best not rebuild the protobuf generated code unless you know something has changed.

BREAKING CHANGE: This module now outputs ESM code only

### CODE REVIEW

**Comment by dapplion** [comment]
Thank you so much! before merging this Lodestar must switch to ESM too, so we'll wait for that to happen

**Comment by achingbrain** [comment]
I'm not sure it's necessary to wait, this can (and should) be published as a major, Lodestar can use the previous version until it's ready to upgrade?

**Review by dapplion — APPROVED** [review]
Looks good to me! @achingbrain yeah you are right we can merge this and keep using the current version is Lodestar. And backport fixes if necessary.

**Review by mpetrunic — APPROVED** [review]

File: src/crypto.ts [hunk 1]
```diff
@@ -1,12 +1,9 @@
-import { IHandshake } from './@types/handshake-interface'
-import { NOISE_MSG_MAX_LENGTH_BYTES, NOISE_MSG_MAX_LENGTH_BYTES_WITHOUT_TAG } from './constants'
-
-interface IReturnEncryptionWrapper {
-  (source: Iterable<Uint8Array>): AsyncIterableIterator<Uint8Array>
-}
+import type { Transform } from 'it-stream-types'
+import type { IHandshake } from './@types/handshake-interface.js'
```
[the remark this answers is not in the document]
Comment (achingbrain) [reply src/crypto.ts:2]: From the way I understand it, `.d.ts` files are meant to be machine generated so they aren't parsed and validated like regular `.ts` files.

We've certainly shipped a lot of broken types in the past due to hand-rolled `.d.ts` files - things like import paths being wrong etc don't trigger errors so it's a lot safer this way.

**Comment by achingbrain** [comment]
Will this get a release soon?  It's the last piece of the puzzle before I can open the ts PR for libp2p itself.

**Comment by mpetrunic** [comment]
will do now

**Comment by mpetrunic** [comment]
> Will this get a release soon? It's the last piece of the puzzle before I can open the ts PR for libp2p itself.

https://github.com/ChainSafe/js-libp2p-noise/releases/tag/v6.0.0

**Pull request merged**